### PR TITLE
Communication board api for school admin

### PR DESCRIPTION
**Pull request by jessyoon14**
학교 관리자를 위한 소통게시판 API를 생성하였습니다.

주기능은, 모든 communication article에 대한 list method, get method, update method입니다.
update method를 사용하여 학교측 관리자가 특정 소통 게시글을 언제 확인했는지 표시할 수 있습니다.

또한, 이 과정에서 Permission 관리를 정리하기 위해 기존에 user_profile의 is_school_admin, is_official을 없애고, user_profile.group를 확장하는 방식으로 통일성을 높였습니다.

테스트 케이스 또한 작성했습니다.

### CODE REVIEW

File: tests/test_communication_article.py [hunk 1]
```diff
@@ -308,3 +343,54 @@ def test_anonymous_comment(self):
...
+        assert res.status_code == 403
+
+    def test_user_cannot_view_admin_api_get(self):
+        res = self.http_request(self.user, 'get', f'communication_articles/{self.article.id}')
+        assert res.status_code == 403
+
+    def test_user_cannot_update_admin_api(self):
+        res = self.http_request(self.user, 'put', f'communication_articles/{self.article.id}', {
+            'confirmed_by_school_at': timezone.now(),
+        })
```
Comment (injoonH) [inline tests/test_communication_article.py:383]: `364` - `366`번째 줄과 같습니다.

File: tests/test_communication_article.py [hunk 2]
```diff
@@ -244,12 +263,20 @@ def test_PREPARING_ANSWER_to_ANSWER_DONE(self):
...
+        from_status = SchoolResponseStatus.PREPARING_ANSWER
+        to_status = SchoolResponseStatus.PREPARING_ANSWER
+
+        article = self._create_article_with_status(from_status)
+        assert self._get_communication_article_status(article) == from_status
+
+        # 정책상 소통게시판에서 vote_cancel은 허용되지 않으나, 테스트 정확성을 위해 첨부함
+        self.http_request(self.user4, 'post', f'articles/{article.id}/vote_cancel')
+        self.http_request(self.user4, 'post', f'articles/{article.id}/vote_positive')
+        assert self._get_communication_article_status(article) == to_status
```
[the remark this answers is not in the document]
Comment (jessyoon14) [reply tests/test_communication_article.py:278]: Vote 수가 30 -> 29 -> 30이 되는 경우를 확인하고자 했습니다.
vote cancel/vote를 user4가 아닌, 다른 user에 대해 진행하면 될 것 같습니다.

File: tests/test_communication_article.py [hunk 3]
```diff
@@ -308,3 +343,54 @@ def test_anonymous_comment(self):
...
+        self.communication_article.refresh_from_db()
+        assert res.status_code == 200
+        assert res.data.get('article') == self.communication_article.article.id
+        assert res.data.get('school_response_status') == self.communication_article.school_response_status
+
+    def test_admin_can_update_admin_api(self):
+        current_time = timezone.now()
+        res = self.http_request(self.school_admin, 'put', f'communication_articles/{self.article.id}', {
+            'confirmed_by_school_at': current_time,
+        })
```
[the remark this answers is not in the document]
Comment (jessyoon14) [reply tests/test_communication_article.py:366]: Admin API의 update에 대한 테스트입니다.
update할 때 시간이 의도한대로 바뀌는지 확인하기 위해, 비교값이 필요합니다. 
그래서 current_time을 따로 변수로 빼주었습니다.
사실 current_time을 안 넣어주면 바로 현재 시간이 들어가겠지만, 그러면 잘 업데이트되었는지 확인할 시간 비교값이 없어지게 됩니다. (min_time 아닌걸로만 확인할 수도 있겠지만, 현재 시각으로 확인하는 것이 더 좋다고 생각했어요)
그런데 다시 생각해보니, 이부분은 min_time에 대한 비교로만 구현해도 괜찮을 것 같습니다.

**Review by jessyoon14 — COMMENTED** [review]
몇가지 답변드렸습니다~

File: tests/test_communication_article.py [hunk 4]
```diff
@@ -130,13 +133,12 @@ def _create_article_with_status(self, status=SchoolResponseStatus.BEFORE_UPVOTE_
             self._add_upvotes(article, users_tuple)
         elif status == SchoolResponseStatus.PREPARING_ANSWER:
             # 작성일 기준 status 변경 방법이 존재하지 않음
```
Comment (injoonH) [inline tests/test_communication_article.py:135]: 주석 삭제하셔도 됩니다.

File: tests/test_communication_article.py [hunk 5]
```diff
@@ -130,13 +133,12 @@ def _create_article_with_status(self, status=SchoolResponseStatus.BEFORE_UPVOTE_
...
-            article.communication_article.response_deadline = timezone.now()
-            article.communication_article.confirmed_by_school_at = timezone.now()
-            article.communication_article.school_response_status = SchoolResponseStatus.PREPARING_ANSWER
-            article.communication_article.save()
+            users_tuple = (self.user2, self.user3, self.user4)
+            self._add_upvotes(article, users_tuple)
```
Comment (injoonH) [inline tests/test_communication_article.py:137]: Threshold에 도달하기 전에 학교 관리자가 `확인했습니다` 버튼 누른 경우를 고려할 수 없습니다.

File: tests/test_communication_article.py [hunk 6]
```diff
@@ -213,9 +215,7 @@ def test_BEFORE_UPVOTE_THRESHOLD_to_PREPARING_ANSWER(self):
         article = self._create_article_with_status(from_status)
         assert self._get_communication_article_status(article) == from_status
 
-        self.http_request(self.school_admin, 'put', f'communication_articles/{article.id}', {
-            'confirmed_by_school_at': timezone.now(),
-        })
+        self.http_request(self.school_admin, 'put', f'communication_articles/{article.id}')
```
Comment (injoonH) [inline tests/test_communication_article.py:218]: Helper function을 사용해주세요.

File: tests/test_communication_article.py [hunk 7]
```diff
@@ -237,9 +237,7 @@ def test_BEFORE_SCHOOL_CONFIRM_to_PREPARING_ANSWER(self):
...
```
Comment (injoonH) [inline tests/test_communication_article.py:240]: `218`번째 줄과 동일합니다.

File: tests/test_communication_article.py [hunk 8]
```diff
@@ -264,18 +262,17 @@ def test_PREPARING_ANSWER_to_ANSWER_DONE(self):
...
-        to_status = SchoolResponseStatus.PREPARING_ANSWER
+        status = SchoolResponseStatus.PREPARING_ANSWER
 
-        article = self._create_article_with_status(from_status)
-        assert self._get_communication_article_status(article) == from_status
+        article = self._create_article_with_status(status)
+        assert self._get_communication_article_status(article) == status
 
         # 정책상 소통게시판에서 vote_cancel은 허용되지 않으나, 테스트 정확성을 위해 첨부함
         self.http_request(self.user4, 'post', f'articles/{article.id}/vote_cancel')
```
Comment (injoonH) [inline tests/test_communication_article.py:273]: `272`번째 줄과 `273`번째 줄 사이에 `self.user4`가 upvote하는 줄을 추가해주세요.

File: tests/test_communication_article.py [hunk 9]
```diff
@@ -287,24 +284,28 @@ def test_ANSWER_DONE_to_BEFORE_SCHOOL_CONFIRM(self):
...
-        assert self._get_communication_article_status(article) == to_status
+        self._confirm_communication_article(article)
+        assert self._get_communication_article_status(article) == status
+
+    def test_days_left(self):
+        status = SchoolResponseStatus.PREPARING_ANSWER
+        article = self._create_article_with_status(status)
+
+        res = self.http_request(self.school_admin, 'get', f'communication_articles/{article.id}')
+        assert res.data.get('days_left') == ANSWER_PERIOD
```
Comment (injoonH) [inline tests/test_communication_article.py:303]: 1. `days_left`가 설정되지 않은 경우도 확인해야 합니다.
2. (**중요**) `UTC`와 `LocalTime` edge case를 고려하는 코드도 추가하면 좋겠습니다.
   - https://stackoverflow.com/a/21677119/13954425 참고하시면 될 것 같습니다.

File: tests/test_communication_article.py [hunk 10]
```diff
@@ -354,43 +355,24 @@ def test_admin_can_view_admin_api_list(self):
...
         assert res.status_code == 200
         assert res.data.get('article') == self.communication_article.article.id
-        assert res.data.get('school_response_status') == self.communication_article.school_response_status
 
     def test_admin_can_update_admin_api(self):
-        current_time = timezone.now()
-        res = self.http_request(self.school_admin, 'put', f'communication_articles/{self.article.id}', {
-            'confirmed_by_school_at': current_time,
-        })
+        res = self.http_request(self.school_admin, 'put', f'communication_articles/{self.article.id}')
```
Comment (injoonH) [inline tests/test_communication_article.py:362]: Helper function 사용해 주세요.

File: tests/test_communication_article.py (hunk 9 above)
Comment (injoonH) [inline tests/test_communication_article.py:303]: 1. `days_left`가 설정되지 않은 경우도 고려해 주세요.
2. (**중요**) `UTC`, `LocalTime` edge case 고려하는 테스트 추가해 주세요.
   - https://stackoverflow.com/a/21677119/13954425 참고하시면 될 것 같습니다.

File: apps/core/serializers/article.py [hunk 11]
```diff
@@ -346,7 +346,7 @@ class ArticleAttachmentType(Enum):
     BOTH = 'BOTH'
 
 
-class ArticleListActionSerializer(HiddenSerializerFieldMixin, BaseArticleSerializer):
+class  ArticleListActionSerializer(HiddenSerializerFieldMixin, BaseArticleSerializer):
```
Comment (injoonH) [inline apps/core/serializers/article.py:349]: Space 하나 제거해 주세요.

File: tests/test_communication_article.py [hunk 12]
```diff
@@ -269,6 +270,12 @@ def test_PREPARING_ANSWER_to_BEFORE_SCHOOL_CONFIRM(self):
         article = self._create_article_with_status(status)
         assert self._get_communication_article_status(article) == status
 
+        # 좋아요 수가 threshold를 넘었을 때
+        for user in (self.user2, self.user3, self.user4):
+            self.http_request(user, 'post', f'articles/{article.id}/vote_positive')
```
Comment (injoonH) [inline tests/test_communication_article.py:275]: `_add_upvotes` 메소드 사용하는 게 좋습니다.

**Review by injoonH — APPROVED** [review]
👍 수고하셨습니다.
